src/main/java/org/openstreetmap/atlas/geography/atlas/change/FeatureChange.java

```diff
@@ -246,131 +278,242 @@ public String toJson()
     {
         return "FeatureChange [changeType=" + this.changeType + ", reference={"
-                + this.reference.getType() + "," + this.reference.getIdentifier() + "}, tags="
+                + this.updatedView.getType() + "," + this.updatedView.getIdentifier() + "}, tags="
                 + getTags() + ", bounds=" + bounds() + "]";
     }
 
-    private FeatureChange mergeAreas(final FeatureChange other,
-            final Map<String, String> mergedTags, final Set<Long> mergedParentRelations)
+    /**
+     * Specify the Atlas on which this {@link FeatureChange} is based. {@link FeatureChange} objects
+     * with a contextual Atlas are able to calculate their before view, and so are able to leverage
+     * richer and more robust merging mechanics.
+     *
+     * @param atlas
+     *            the contextual atlas
+     * @return the updated {@link FeatureChange}
+     */
+    public FeatureChange withAtlasContext(final Atlas atlas)
     {
-        final AtlasEntity thisReference = this.getReference();
-        final AtlasEntity thatReference = other.getReference();
-        final Polygon mergedPolygon = mergedMember("polygon", thisReference, thatReference,
-                atlasEntity -> ((Area) atlasEntity).asPolygon(), Optional.empty());
-
-        CompleteArea result = new CompleteArea(getIdentifier(), mergedPolygon, mergedTags,
-                mergedParentRelations);
-        if (result.bounds() == null)
+        // TODO before view for REMOVE type?
+        if (this.changeType == ChangeType.ADD)
         {
-            if (bounds() != null)
-            {
-                result = result.withAggregateBoundsExtendedUsing(bounds());
-            }
-            else if (other.bounds() != null)
-            {
-                result = result.withAggregateBoundsExtendedUsing(other.bounds());
-            }
+            computeBeforeViewUsingAtlasContext(atlas);
         }
-        return FeatureChange.add(result);
+        return this;
     }
 
-    private <M> M mergedMember(final String memberName, final AtlasEntity left,
-            final AtlasEntity right, final Function<AtlasEntity, M> memberExtractor,
-            final Optional<BinaryOperator<M>> memberMergerOption)
+    /**
+     * Compute the beforeView using a given afterView and Atlas context. The beforeView is a
+     * CompleteEntity, and will contain only those fields which have been updated in the afterView.
+     * E.g. Suppose the afterView is a CompleteNode with ID 1, and the only updated field is the tag
+     * Map. Then the beforeView will be a CompleteNode with ID 1, where the only populated field is
+     * the tag Map. However, the afterView's tag Map will be fetched from the Atlas context. Having
+     * a beforeView for each {@link FeatureChange} allows for more robust merging strategies.
+     *
+     * @param atlas
+     *            the atlas context
+     */
+    private void computeBeforeViewUsingAtlasContext(final Atlas atlas) // NOSONAR
```

Comment:
Here is how we compute the beforeView of a `FeatureChange` based on a given atlas context. This beforeView will be a `CompleteEntity`, and only contain the fields that have changed.